NeoML/include/NeoML/Dnn/Dnn.h

```diff
@@ -700,6 +686,56 @@ inline CArchive& operator>>( CArchive& archive, CDnn& dnn)
 
 void NEOML_API SerializeLayer( CArchive& archive, IMathEngine& mathEngine, CPtr<CBaseLayer>& layer );
 
+//------------------------------------------------------------------------------------------------------------
+
+// Result of CReferenceDnnFactory::CreateReferenceDnn
+class NEOML_API CDnnReference : public IObject {
+public:
+	CDnn Dnn;
+
+	static CPtr<CDnnReference> Create( CRandom& random, IMathEngine& mathEngine )
+	{ return new CDnnReference( random, mathEngine ); }
+
+private:
+	CDnnReference( CRandom& random, IMathEngine& mathEngine ) : Dnn( random, mathEngine ) {}
+	friend class CReferenceDnnFactory;
+};
+
+// This class can initialize a reference dnn, that has the same configuration as the original dnn
+// and shares parameter blobs with the original dnn to save memory.
+// Useful for multi-threaded inference where each thread can operate own reference dnn independently.
+// Learning is disabled for both the original dnn and the reference dnn.
+// Creates a copy of the original dnn's random generator to use it for inference.
+class NEOML_API CReferenceDnnFactory : public IObject {
+public:
+	// Archive should contain trained dnn, ready for inference
+	// NOTE: mathEngine should be CPU only and live longer than CReferenceDnnFactory
+	static CPtr<CReferenceDnnFactory> Create( IMathEngine& mathEngine, CArchive& archive,
+		int seed = 42, bool optimizeDnn = true );
+	// Dnn should be trained and ready for inference,
+	// Dnn will be copied for internal storage to be non-disturbed, one can use argument dnn further as one wants
+	// NOTE: mathEngine should be CPU only and live longer than CReferenceDnnFactory
+	static CPtr<CReferenceDnnFactory> Create( IMathEngine& mathEngine, const CDnn& dnn, bool optimizeDnn = true );
+	// Dnn should be trained and ready for inference, it will be moved inside and cannot be used outside.
+	// NOTE: mathEngine should be CPU only and live longer than CReferenceDnnFactory
+	static CPtr<CReferenceDnnFactory> Create( CDnn&& dnn, bool optimizeDnn = true );
+
+	// Thread-safe coping of originalDnn, increments the counter
+	// NOTE: The original dnn used to copy reference dnns may be also used as one more reference dnn (optimization)
+	CPtr<CDnnReference> CreateReferenceDnn( bool getOriginDnn = false );
```

Comment:
I think - now it is an ideal version.

No need to take an attention to `CPtr<CReferenceDnnFactory>`, it will be destroyed with the last reference dnn destruction
There it no CPointers loops, because the origin dnn is not need the factory

No need to take an attention to `CPtr<CDnnReference>`, there is not difference with reference and origin dnn.
But if one wants to does not use the origin dnn for optimization, one can it do.
`CDnnReference` is not in `CDnn` hierarchy.

Added documentation

Reply:
Minor thing, but the flag makes a behavior very confusing. Normally the method creates a **new** reference ("soft copy") which can be safely used in a new thread, but with flag it just returns the **same** object every time. It fact, it is a completely different method inside the other one. 

I argued against the method GetOriginalDnn() because it fixes the implementation details of having such a Dnn and the possibility of its public use. Probably it also had different semantics. In this version it does all the same and confuses things on top of that. Summing up, if you really want to make originalDnn public, please make a separate method. 

I also suggested to return originalDnn at first call of CreateReferenceDnn() instead of creating a reference. There should be no difference for user between original and reference. If there is some difference, don't show the original. Don't make it a swiss knife that can cut one's finger (or throat)